pandas/core/arrays/datetimes.py

```diff
@@ -458,72 +467,27 @@ def _add_offset(self, offset):
-        elif isinstance(other, (datetime, np.datetime64)):
-            assert other is not NaT
-            other = Timestamp(other)
-            if other is NaT:
-                return self - NaT
+        assert isinstance(other, (datetime, np.datetime64))
+        assert other is not NaT
+        other = Timestamp(other)
+        if other is NaT:
+            return self - NaT
+
+        if not self._has_same_tz(other):
             # require tz compat
-            elif not self._has_same_tz(other):
-                raise TypeError("Timestamp subtraction must have the same "
-                                "timezones or no timezones")
-            else:
-                i8 = self.asi8
-                result = checked_add_with_arr(i8, -other.value,
-                                              arr_mask=self._isnan)
-                result = self._maybe_mask_results(result,
-                                                  fill_value=iNaT)
-        else:
-            raise TypeError("cannot subtract {cls} and {typ}"
-                            .format(cls=type(self).__name__,
-                                    typ=type(other).__name__))
+            raise TypeError("Timestamp subtraction must have the same "
+                            "timezones or no timezones")
+
+        i8 = self.asi8
+        result = checked_add_with_arr(i8, -other.value,
+                                      arr_mask=self._isnan)
+        result = self._maybe_mask_results(result, fill_value=iNaT)
         return result.view('timedelta64[ns]')
 
+    @Appender(dtl.DatetimeLikeArrayMixin._add_delta.__doc__)
     def _add_delta(self, delta):
-        """
-        Add a timedelta-like, DateOffset, or TimedeltaIndex-like object
-        to self.
-
-        Parameters
-        ----------
-        delta : {timedelta, np.timedelta64, DateOffset,
-                 TimedeltaIndex, ndarray[timedelta64]}
-
-        Returns
-        -------
-        result : same type as self
-
-        Notes
-        -----
-        The result's name is set outside of _add_delta by the calling
-        method (__add__ or __sub__)
-        """
-        from pandas.core.arrays import TimedeltaArrayMixin
-
-        if isinstance(delta, (Tick, timedelta, np.timedelta64)):
-            new_values = self._add_delta_td(delta)
-        elif is_timedelta64_dtype(delta):
-            if not isinstance(delta, TimedeltaArrayMixin):
-                delta = TimedeltaArrayMixin(delta)
-            new_values = self._add_delta_tdi(delta)
-        else:
-            new_values = self.astype('O') + delta
-
-        tz = 'UTC' if self.tz is not None else None
-        result = type(self)(new_values, tz=tz, freq='infer')
-        if self.tz is not None and self.tz is not utc:
-            result = result.tz_convert(self.tz)
-        return result
+        new_values = dtl.DatetimeLikeArrayMixin._add_delta(self, delta)
+        return type(self)(new_values, tz=self.tz, freq='infer')
```

Comment:
Much less verbose than the old 522-526.  Not sure if there was a reason for the old version.

Reply:
Why does this use `type(self)(new_values)`, while the PeriodArrayMixin version uses `self._shallow_copy`? Is that necessary due to inconsistent constructors, or can we standardize?

Reply:
Because shallow_copy doesn't allow `freq='infer'`

Reply:
@jbrockmendel related to your initial comment. If I do the following (mimicking the old and new code):

```
In [15]: new_values = np.array([1, 2, 3], dtype='datetime64[ns]')

In [16]: pd.DatetimeIndex(new_values, tz='Europe/Brussels')
Out[16]: 
DatetimeIndex(['1970-01-01 00:00:00.000000001+01:00',
               '1970-01-01 00:00:00.000000002+01:00',
               '1970-01-01 00:00:00.000000003+01:00'],
              dtype='datetime64[ns, Europe/Brussels]', freq=None)

In [17]: pd.DatetimeIndex(new_values, tz='UTC').tz_convert('Europe/Brussels')
Out[17]: 
DatetimeIndex(['1970-01-01 01:00:00.000000001+01:00',
               '1970-01-01 01:00:00.000000002+01:00',
               '1970-01-01 01:00:00.000000003+01:00'],
              dtype='datetime64[ns, Europe/Brussels]', freq=None)
```

is not giving the same result (or did I do something wrong in the example? I assumed that `self.tz` was 'Europe/Brussels')

But that is just from looking at the small piece of code here in the diff, not the broader picture where and how it is called etc.

Reply:
@jorisvandenbossche turns out I had it right the first time.  The problem with the example you gave is that you're passing `M8[ns]` instead of `i8`.

```
In [3]: new_values = np.array([1, 2, 3])

In [4]: pd.DatetimeIndex(new_values, tz='Europe/Brussels')
Out[4]: 
DatetimeIndex(['1970-01-01 01:00:00.000000001+01:00',
               '1970-01-01 01:00:00.000000002+01:00',
               '1970-01-01 01:00:00.000000003+01:00'],
              dtype='datetime64[ns, Europe/Brussels]', freq=None)

In [5]: pd.DatetimeIndex(new_values, tz='UTC').tz_convert('Europe/Brussels')
Out[5]: 
DatetimeIndex(['1970-01-01 01:00:00.000000001+01:00',
               '1970-01-01 01:00:00.000000002+01:00',
               '1970-01-01 01:00:00.000000003+01:00'],
              dtype='datetime64[ns, Europe/Brussels]', freq=None)
```

Reply:
Ah ... subtle difference :) 
Then the docstring of `_add_delta` is incorrect I think.

Reply:
> Then the docstring of _add_delta is incorrect I think.

Can you elaborate on this?

Reply:
> Can you elaborate on this?

Added a comment on the relevant line

Reply:
> Can you, regardless of the docstring appender, add a single line comment like # add a timedelta-like or DateOffset, yielding a ... (as is done for some of the others as well).
That at least makes reading the code a bit easier.

Can you do this one?

Reply:
Can you do this one?

Reply:
Just updated.  It's much less pretty than before, but more Technically Correct.